### PR TITLE
Ruby: Add support for GraphQL

### PR DESCRIPTION
**Pull request by jeffgran**
Adds support for [GraphQL Ruby](https://graphql-ruby.org/).

This is the most widely-used GraphQL library in the ruby ecosystem. It integrates with Rails but all of the handling happens in the GraphQL classes (and subclasses), completely disconnected from the Rails Controllers, etc.

What I have done here is added the CodeQL classes necessary to expose the methods that handle user input as `HTTP::Server::RequestHandler::Range` subclasses. That way, all the existing path queries that deal with user input (RemoteFlowSource) will work in GraphQL handlers as well.

This was my first time writing CodeQL - I'm sure I did some wonky things. Help on improving this code is welcome.

### CODE REVIEW

**Comment by hmac** [comment]
Hi @jeffgran - sorry for the slow response to this PR, and thank you for opening it! This looks like a great addition to our library modelling. I will follow up later today with a more detailed review, but wanted to let you know that we're taking a look.

There are a few WIP commits on this branch - have you reached a finished state or are you still working on this in places?

**Comment by jeffgran** [comment]
Hi @hmac - Yeah, there are two `TODO` comments:

1. Adding support for the graphql `Interface` - I think I could add that with similar techniques to what I've done already. I wanted to see whether you'd be amenable to merging this work before I spent more time on that. 
2. A potential false positive -- I don't know how to solve this, but it seems like more of a generic ruby issue than specific to this PR. It's that a specific keyword arg could be a source of untrusted input, but if the method takes double-splatted `**kwargs`, then I don't know how to target just the specific key of that hash in the method. Maybe you know of a generic way to handle this?

The other thing I think would be good, before this is merged, is to add more extensive documentation for all the classes I wrote. Happy to do that, but again I wanted to see whether you'd consider this as an addition before I spent time on that polish.

Thanks for the response. Hoping we can get this in!

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:8]: We don't (yet) want users to rely on these classes, as they may change in the future. So let's wrap everything in a `private module`.

```ql
private module GraphQL {
  ...
}
```

We can then drop the `private` annotations from the individual classes.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:14]: We can use API Graphs for these kinds of queries. We don't yet have documentation for the Ruby API Graphs library, but you can replace the above with
```ql
private import codeql.ruby.ApiGraphs

...

this = API::getTopLevelMember("GraphQL")
         .getMember("Schema")
         .getMember("RelayClassicMutation")
         .getAUse()
         .asExpr()
         .getExpr()
```

If you want you could then go futher and inline this class into its only use, in `GraphqlRelayClassicMutationClass`.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:23]: You can use API Graphs here too.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:32]: Ditto here.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:110]: As you mentioned already, having documentation for each of these classes would be great.

**Review by hmac — COMMENTED** [review]
@jeffgran that makes sense. I've made an initial pass with some suggestions for improvements. I'm not super familiar with GraphQL so having documentation on each class would help me to review it in more detail 😄 

Your point about `**kwargs` sounds like a limitation with our dataflow modelling at the moment. We're working on improving dataflow through collections, so we might be able to fix this particular issue in the future, but for now it may be the best we can do.

If you haven't already, you'll want to format the QL code with `codeql query format`.

**Comment by jeffgran** [comment]
Hi @hmac , thanks for working with me on this!

I just pushed https://github.com/github/codeql/pull/7126/commits/9e2ca7a7b5c6f05cac2f373298bfcfc923c80250 , in which I:
- switched those `Access` classes to use the API Graphs library
- added docstrings to all the classes i wrote
- formatted the code using the auto-format cli command
- added a private `GraphQL` module

I don't think I quite understood your suggestion to wrap everything in a private module though. I tried wrapping the entire library I wrote in a private module, but then my tests stopped working because I couldn't import/resolve those classes... which I think means that my addition [here](https://github.com/github/codeql/pull/7126/files?diff=unified&w=0#diff-759b362be9329c2cd7090fbee45317ee631c25807da29d2a99fba0d307bec91aR9) wouldn't work either. But that's the "magic" that makes this addition automatically detect any user-input type vulnerabilities because of my extensions of `HTTP::Server::RequestHandler::Range`. Let me know if I misunderstood, or if there's a way to make this library code private but still use import it in other files?

Also just to note, I removed the TODO comment about adding support for the GraphQL `Interface` type. I do think it would be a useful addition, and I think there are even a few more ways besides that, that user input can get into code through the graphql library. But I think what I have here covers the most common usage of the library so would still be beneficial as-is, as a first pass. Would you like to hold off merging this until it covers all possible vectors for user input, or would implementing this in phases be ok? I'll work with whichever you prefer.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:295]: There's a convention (for some reason) that getters like this have a doc comment that starts `Gets the ...`. Unfortunately I think the CI job that catches this doesn't run on external PRs. We will want to make the same change to the other cases in this file.

```suggestion
  /** Gets the `GraphqlFieldableClass` representing the receiver of this method. */
```

**Comment by hmac** [comment]
Nice work!

On the `private module` thing, I think I gave you some bad advice there, sorry. As you say, in order to test these classes/predicates they need to be exposed. Let's remove the module and just make `private` anything that doesn't need to be public (i.e. anything that isn't mentioned in `GraphQL.ql`). The intention is to reduce the API surface area so that we minimise breakage to consumers if we change these classes in the future. I'll double check with the team to see if there's anything else we want to do on this front.

> Would you like to hold off merging this until it covers all possible vectors for user input, or would implementing this in phases be ok?

Implementing in phases sounds like a good plan! Happy to keep the scope of this PR small and we can make further improvements in followups :)

File: ruby/ql/test/library-tests/frameworks/GraphQL.ql
Comment (hmac) [inline ruby/ql/test/library-tests/frameworks/GraphQL.ql:1]: This file needs running through `codeql query format`, I think.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:271]: Is there a reason this class exists, separate from `GraphqlSchemaObjectClass`? They seem like fundamentally the same thing, except this has the extra member predicate `getAFieldDefinitionMethodCall`. If possible, it would be nice to merge the two together - I feel that `GraphqlSchemaObjectClass` is a better, more self-descriptive name.

Comment (jeffgran) [reply ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:271]: Yes, this class exists because it's where I would eventually like to add support for the [`GraphQL::Schema::Interface`](https://graphql-ruby.org/type_definitions/interfaces.html#definition-methods) functionality, where you can create reusable modules that contain a common set of fields. However, I'm not sure exactly how I'll implement that, and since those are ruby `module`s, not `class`es, It likely wouldn't fit in here anyway. Maybe I'd be able to use like a `union` type or something. Anyway, good call-out. I'll remove this for now.

**Comment by jeffgran** [comment]
@hmac I believe I addressed all your concerns with https://github.com/github/codeql/pull/7126/commits/bca5adaf84865085cb3a782ea6fef8b2bc21a8a9 - let me know if you have any more. I think we're getting pretty close here!

**Comment by hmac** [comment]
@jeffgran sorry again for the delay! I think there's just one small linting error and this should be good to go. If you rebase this PR then that check should hopefully stop complaining about "analysis not found" - I think the config that enables it was added after this branch was forked off `main`.

**Comment by jeffgran** [comment]
@hmac no worries at all, we're all busy :) Just rebased and removed that redundant cast. 

🤞🏻

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:407]: From looking at the graphql-ruby [docs](https://graphql-ruby.org/fields/arguments), it looks like Graphql arguments are always passed as keyword arguments, so should we be using `KeywordParameter` here?

If so, that means we can drop line 409 as `KeywordParameter` and `HashSplatParameter` are disjoint.

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:414]: This is probably _slightly_ more idiomatic as `result instanceof HashSplatParameter`.

File: ruby/ql/test/library-tests/frameworks/app/graphql/types/query_type.rb
Comment (hmac) [inline ruby/ql/test/library-tests/frameworks/app/graphql/types/query_type.rb:30]: I found it useful to add a test here for a method that captures one of its arguments as a kwarg and the other in a hash splat, to check that both were identified as remote flow sources - that might be worth adding? Something like
```ruby
field :with_splat_and_named_arg, String, null: false, description: "A field with two named arguments, where the method captures the second via a hash splat param" do
  argument :arg1, Int, "A number", required: true
  argument :arg2, Int, "Another number", required: true
end
def with_splat_and_named_arg(arg1:, **rest)
  system("echo #{arg1}")
  system("echo #{rest[:arg2]}")
  arg1.to_s
end
```

File: ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll
Comment (jeffgran) [reply ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll:407]: yep, makes sense.

**Comment by hmac** [comment]
@jeffgran I'm actually not sure why the tests are failing here - they seem to pass when I run them locally. Perhaps if you rebase the PR against `main` then the problem might become clearer (or they might magically start passing 🤞)?

**Comment by jeffgran** [comment]
@hmac rebased, 🤞🏻 here too :)

**Comment by hmac** [comment]
Tests pass 🎉 

@jeffgran if you squash your commits down and add an informative commit message then we can merge this 👍

**Comment by jeffgran** [comment]
@hmac done!

**Review by hmac — DISMISSED** [review]

**Review by hmac — APPROVED** [review]

**Comment by hmac** [comment]
...and done. Thanks for you work (and patience) @jeffgran!

**Comment by jeffgran** [comment]
Thanks so much for working with me on this too, probably the biggest OSS contribution I've ever made :)

**Pull request merged**
